### ISSUE TITLE
Http Backlog

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#2793]
Prioritized list of features and issues to take care of relating to the `Http` library.

## Closed
 - [x] #2816 _chore(http): add XMLHttpRequest as type on XHRConnection_
 - [x] #2810 _feat(http): simple, dynamic retrying of failed requests_

## Implemented

 - [x] #2564 _refactor(Http): remove HttpFactory_
 - [x] #2415 _feat(http): EventEmitter and Dart support_
 - [x] #2818 _feat(http): add JsonP and JsonPBackend_
 - [x] #2680 _refactor(http): move http into its own module as sibling of angular2_
 - [x] #3365 _fix(http): cannot use jsonpInjectables and httpInjectables in same injector_
 - [x] #2417 _feat(http): implement support for headers and url params objects in http_
 - [x] #2635 _http should complete after response_
 - [x] #2904 _chore(http): convert RequestMethods to const list_
 - [x] #3110 _feat(Observable): swap underlying RxJS implementation used by EventEmitter with RxJS-Next_ (Blocked by RxJS alpha release)
 - [x] #3339 _chore(http): remove unused properties from Request and Response_
 - [x] #2788 _Http class requires `@Inject` decorator to be injected when using TypeScript_
 - [x] #4245 _fix(http): should throw an exception if input to `http.request` is not string or `Request`_
 - [x] #2800 _feat(http): implement error propagation in XHRBackend_
 - [x] #4043 _Http requests should be lazy / cold Observables_
 - [x] #2840 _feat(http): accept string for method in RequestOptions class and interface_
 - [x] #2794 @robwormald  M38  _feat(http): directly return RxJS Observable or Dart Stream from Http 
methods_
 - [x] #4588 _Fix async pipe so it works with Http / proper Observables_
 - [x] #5015 Fix Connection.response typedef to Observable<Response>

## Blocked

 - [ ] #2684 Needs design _RFC: Http interceptors and transformers_
 - [ ] #2822 Needs design _feat(http): helpers/guides for short polling_
 - [ ] Security features (No issue yet, needs design)

## Ready to Implement

Ordered roughly by a factor of priority & level of effort
 - [ ] #2554 _refactor(http): move mock_backend to test_lib module_
 - [ ] #4472 _refactor(http): convert Connection interface to be Observable of Response_
 - [ ] #4231 _feat(http): add support for withCredentials_
 - [ ] #2897 _feat(http): support use by AngularJS 1.x applications_
 - [ ] #2806 _chore(http): assert that url is present in RequestOptions when creating a Request_
 - [ ] #2609 _perf(http): Provide means to optimize response parsing_
 - [ ] #2803 _feat(http): Support many types of body on Request/Response and all ConnectionBackend implementations_
 - [ ] #2903 _feat(http): automatically set request Content-Type header based on body type_
 - [ ] #2812 _feat(http): add timeout support to `Request`_
 - [ ] #2795 _feat(http): implement streaming upload/download_
 - [ ] #2799 _refactor(http): only use readyState in mock connection_
 - [ ] #3513 _refactor(http): rename URLSearchParams to something else_

**Comment by jeffbcross** [comment]
@mhevery this issue is a meta-issue created only because of Github's deficiencies in organizing and prioritizing issues. This represents all open http-related issues, to be completed over many milestones, in order of priority. None of our labels adequately capture its purpose, and I don't want the issue to be lumped in with other "feature" issues when filtering by labels. 

It'd be nice to have a "type: backlog" label, which would keep issues like this separate from feature/bug issues in different views, as well as keep it from showing up as "untriaged." @naomiblack does this sound reasonable?

**Comment by mhevery** [comment]
Backlog sounds reasonable. But isn't that just `P!: Backlog`? (just create such a priority and it should be all you need.)

**Comment by jeffbcross** [comment]
I bumped up priority of moving http to its own module, since in Dart, the inclusion of the jsonp file, which imports dart:html, causes problems when importing the top-level angular2/angular2 module in a non-browser environment.

**Comment by markharding** [comment]
Is it finalised that Http request responses will be `EventEmitter`? What are the plans in returning other events, such as progress, rather than just a response?

Perhaps it would be better to just return the connection and then backends such as `XHRConnection` can return `response`, `progress`, etc events instead?

**Comment by PascalPrecht** [comment]
https://github.com/angular/angular/pull/3390 can be ticked on that list.

**Comment by jeffbcross** [comment]
Thanks @PascalPrecht! I updated the list status with yours and some other PRs.

**Comment by PascalPrecht** [comment]
#3339 is also done here: https://github.com/angular/angular/pull/3823

(or at least ready for review)

**Comment by robwormald** [comment]
@jeffbcross fyi on this one https://github.com/angular/angular/issues/4043, might affect a number of these

**Comment by jeffbcross** [comment]
I'm closing this issue since we are opening http issues at a high frequency, and there is a special label for http not. Unfortunately, searching by the label doesn't give any real insight into priority of http issues, but it's better than stale information here :) https://github.com/angular/angular/labels/comp%3A%20http

**Comment by Leavesye** [comment]
fdgfdg
